parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageWriteStore.java

```diff
@@ -178,12 +184,23 @@ public long getMemSize() {
       return buf.size();
     }
 
-    public void writeToFileWriter(ParquetFileWriter writer) throws IOException {
+    public List<PageHeaderWithOffset> writeToFileWriter(ParquetFileWriter writer) throws IOException {
+      final List<PageHeaderWithOffset> allPageHeaders = new ArrayList<PageHeaderWithOffset>();
       writer.startColumn(path, totalValueCount, compressor.getCodecName());
       if (dictionaryPage != null) {
-        writer.writeDictionaryPage(dictionaryPage);
+        allPageHeaders.add(writer.writeDictionaryPage(dictionaryPage));
         encodings.add(dictionaryPage.getEncoding());
       }
+
+      // start from current offset in output file, until now page with offsets have saved page sizes.
+      long pageOffset = writer.getPos();
```

Comment:
if you keep the position relative to the 1st datapage above, you just need add this value to every PageHeaderWithOffset here